sdks/java/core/src/main/java/org/apache/beam/sdk/util/RowJson.java

```diff
@@ -527,7 +527,9 @@ private void writeValue(JsonGenerator gen, FieldType type, Object value) throws
           gen.writeNumber((BigDecimal) value);
           break;
         case DATETIME:
-          gen.writeString(((DateTime) value).toString()); // ISO 8601 format
+          assert (value instanceof ReadableInstant);
```

Comment:
I recognize this is effectively a comment, with the added benefit that we'll get a nicer error in tests.

Reply:
nit: If you cast `value` to `ReadableInstant` you get similar behavior but it isn't sometimes a comment. What is the advantage of this way?

Reply:
I wasn't sure it was desirable to fail if value is not a ReadableInstant.

Then again, I think your nudge has convinced me that is desirable, since we generally prefer failing rather than producing an incorrect result.